Scheduler reconnects

lmars opened the pull request:
- Update the scheduler to reconnect to the controller if there is an error or if the channel closes, waiting 1 second if more than one connection attempt has been made since the last update
- Add `ExpandedFormation.UpdatedAt` so that the scheduler can ask for only the updates since the last formation it received an event for
- Make deleted formations appear to have no processes so that the scheduler will remove it's processes if necessary

lmars committed:
Update the scheduler to reconnect to the controller

Signed-off-by: Lewis Marshall <lewis@lmars.net>

titanous commented on formation.go:
```diff
@@ -86,12 +88,12 @@ func scanFormation(s Scanner) (*ct.Formation, error) {
 
 func (r *FormationRepo) List(appID string) ([]*ct.Formation, error) {
-	rows, err := r.db.Query("SELECT app_id, release_id, processes, created_at, updated_at FROM formations WHERE app_id = $1 AND deleted_at IS NULL ORDER BY created_at DESC", appID)
+	rows, err := r.db.Query("SELECT app_id, release_id, processes, created_at, updated_at, deleted_at FROM formations WHERE app_id = $1 AND deleted_at IS NULL ORDER BY created_at DESC", appID)
```
Both of these queries filter `deleted_at IS NULL`, and we also set `processes = NULL` when deleting. There is probably a more efficient way of handling deleted formations in general, does the commit have an impact on the reconnect part?

titanous replied:
Hmm, if we bump `updated_at` when we delete, we should be fine.

titanous replied:
If we keep the `deleted_at IS NULL` filter, we don't need to select `deleted_at`.

lmars replied:
Ah to be honest I missed the fact that removing a formation sets `processes = NULL`. I think as you say we just need to ensure removing a formation sets `updated_at`. I'll make that change

titanous commented on formation.go:
```diff
@@ -119,7 +121,8 @@ func (r *FormationRepo) publish(appID, releaseID string) {
 		// formation delete event
-		formation = &ct.Formation{AppID: appID, ReleaseID: releaseID}
+		updated_at := time.Now()
+		formation = &ct.Formation{AppID: appID, ReleaseID: releaseID, UpdatedAt: &updated_at}
```
Is this a workaround for deleted formations not having the right `updated_at`?

lmars replied:
Well, I want all published formation updates to have an `UpdatedAt` field so the scheduler can always get only the updates it has missed when re-connecting, but here I don't get the deleted formation from `r.Get`, so I am just setting it to something reasonable.

I guess I could add a flag to `r.Get` which will cause it to return the deleted formation and I could then use it's updated_at? WDYT?

titanous replied:
Ah, I see. The current implementation is fine.

titanous commented on formation.go:
```diff
@@ -66,18 +66,20 @@ func (r *FormationRepo) Add(f *ct.Formation) error {
+			n, _ := strconv.Atoi(v.String)
+			if n > 0 {
+				f.Processes[k] = n
+			}
```
If we do this, we can stop setting `processes = NULL`.

titanous replied:
Actually, I just remembered that we set `processes = NULL` on delete to send the `NULL` down the stream to the scheduler. I don't think we need to do this `DeletedAt` check.

lmars replied:
Yep I have dropped this change and just ensured that `updated_at` is set when removing a formation: a140643

titanous commented on scheduler/main.go:
```diff
@@ -165,36 +165,54 @@ func (c *context) syncCluster(events chan<- *host.Event) {
+
+		g.Log(grohl.Data{"at": "connect", "attempt": attempts})
+		ch, err := c.StreamFormations(&lastUpdatedAt)
+		if err != nil {
```
`StreamFormations` returns a `*error`, so this should never be `nil`. Check it after `ch` has closed, and the `range` ends (which will be immediately if the RPC fails).

lmars replied:
Hmm maybe I don't quite understand go pointers, but if `err` is never nil then this would always log an error and continue to the next iteration, which isn't happening.

titanous replied:
Yes, the fake controller client is incorrectly returning a nil `*error`: https://github.com/flynn/flynn-controller/blob/80290445cc8d3c9da33f9e7e08893520286af14f/scheduler/scheduler_test.go#L64

Did you test manually, or just with the test suite?

lmars replied:
Ah yes of course, I did not test against a running controller, updated in a33f3b2

titanous commented on scheduler/main.go:
```diff
@@ -165,36 +165,54 @@ func (c *context) syncCluster(events chan<- *host.Event) {
+			attempts = 0
+
+			if ef.App == nil {
+				// sentinel
```
If we do a full-cluster sync after reconnecting, I'm pretty sure we don't need to worry about deleted formations.

lmars replied:
But what if a formation is deleted whilst the scheduler is disconnected? A full-cluster sync would not rectify that would it?

titanous replied:
Hmm, it should get a 404 when it looks up the formation for a running process from a deleted formation. We should test that and make sure it works as expected.

lmars replied:
So are you assuming when we do another cluster sync we will first reset the schedulers job / formation map? Otherwise there would be no lookup right?

titanous replied:
Yeah, I think that makes sense. That way there will be no difference in behavior between a scheduler process restarting and reconnecting.

titanous replied:
Actually, that mitigates the usefulness of `since`...

titanous replied:
We could send newly deleted formations from the controller when `since` is non-zero.

lmars replied:
Yeh I was thinking that, so should I just drop the updated_at tracking?

lmars replied:
Well, I guess I still need to know if there was a new formation in the time I was disconnected

lmars committed:
Set updated_at when removing a formation

This is so that it will appear as an update when Subscribing to
formation events since a particular time

Signed-off-by: Lewis Marshall <lewis@lmars.net>

titanous commented on formation.go:
```diff
@@ -108,7 +108,7 @@ func (r *FormationRepo) List(appID string) ([]*ct.Formation, error) {
 
 func (r *FormationRepo) Remove(appID, releaseID string) error {
-	err := r.db.Exec("UPDATE formations SET deleted_at = now(), processes = NULL WHERE app_id = $1 AND release_id = $2", appID, releaseID)
+	err := r.db.Exec("UPDATE formations SET deleted_at = now(), processes = NULL, updated_at = now() WHERE app_id = $1 AND release_id = $2", appID, releaseID)
```
Ditto here: https://github.com/flynn/flynn-controller/blob/80290445cc8d3c9da33f9e7e08893520286af14f/app.go#L170

lmars replied:
Done: 15f0c31

lmars committed:
Drop comment from scheduler

Since we track the last update seen from the controller and request all
updates after that when reconnecting, there is no need to sync with the
cluster when reconnecting.

Signed-off-by: Lewis Marshall <lewis@lmars.net>
